### PR TITLE
feat(ngMock.$componentController): add helper to instantiate controllers for components

### LINKED ISSUE #13683: REQUEST FOR FEEDBACK: `angular.component()` - auto creating a controller
As discussed with @shahata and @petebacondarwin - 

The problem I run into when writing lots of "component like" directives is the unnecessarily repeated configuration over and over again.

### It's redundant to have a separate controller declaration.

Registering controllers by name in a component based architecture is useful only for unit tests, where you'd want to use `$controller` to create an instance of the component's controller.

But registering it over and over again is redundant and can be done by the `.component`.

### We should have one `.comp.js` file

If we use `.component()`, it should auto create a new controller, registered with the same component name (because it too must be unique).

So we should end up with one file where the `.component()` becomes a true "annotation like" thing, something like: 

```js
angular
 .module('myModule')
 .component('productList', {
   controller: ProductListController,
   templateUrl: 'product-list.comp.html'
  })

function ProductListController(){

}

$controller('productList') // <-- should be an instance of ProductListController

```

I'd love to hear the cons for this approach

What do you think?

### PR DESCRIPTION
**Pull request by petebacondarwin**
Closes #13683

### CODE REVIEW

File: test/ngMock/angular-mocksSpec.js [hunk 1]
```diff
@@ -1857,6 +1857,65 @@ describe('ngMock', function() {
...
+          controller: 'TestController'
+        });
+      });
+      inject(function($componentController, $rootScope) {
+        var $scope = {};
+        var ctrl = $componentController('test', {$scope: $scope, a: 'A', b: 'B'}, { x: 'X', y: 'Y' });
+        expect(ctrl).toEqual({$scope: $scope, a: 'A', b: 'B', x: 'X', y: 'Y' });
+      });
+    });
+  });
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: I'd add a test with `controller: 'TestController as test'`

File: test/ngMock/angular-mocksSpec.js [hunk 2]
```diff
@@ -1857,6 +1857,65 @@ describe('ngMock', function() {
...
+      }
+      module(function($compileProvider) {
+        $compileProvider.component('test', {
+          controller: TestController
+        });
+      });
+      inject(function($componentController, $rootScope) {
+        var $scope = {};
+        var ctrl = $componentController('test', {$scope: $scope, a: 'A', b: 'B'}, { x: 'X', y: 'Y' });
+        expect(ctrl).toEqual({$scope: $scope, a: 'A', b: 'B', x: 'X', y: 'Y' });
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: Super-nitpicking: Inconsistent spacing before/after `{` / `}` (affects other tests too) :stuck_out_tongue:

<sub>(Sorry, I couldn't help it :flushed:)</sub>

File: src/ngMock/angular-mocks.js [hunk 3]
```diff
@@ -2161,6 +2161,28 @@ angular.mock.$ControllerDecorator = ['$delegate', function($delegate) {
...
+ * @param {string} componentName the name of the component whose controller we want to instantiate
+ * @param {Object} locals Injection locals for Controller.
+ * @param {Object=} bindings Properties to add to the controller before invoking the constructor. This is used
+ *                           to simulate the `bindToController` feature and simplify certain kinds of tests.
+ * @return {Object} Instance of requested controller.
+ */
+angular.mock.$ComponentControllerProvider = ['$compileProvider', function($compileProvider) {
+  return {
+    $get: ['$controller', function($controller) {
+      return function $directiveController(componentName, locals, bindings, ident) {
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: Nitpicking: `$directiveController` is a somewhat misleading name.

**Comment by Narretz** [comment]
Wouldn't it be nice to have that for plain directives, too?

**Comment by gkalpak** [comment]
There will be some slight inconsistency between using `controller: 'SomeCtrl as identifier'` and `controller: 'SomeCtrl'` or `controller: SomeCtrl`. In the former case, the controller will be added as an alias on the scope, whereas in other two cases, it won't (when usign `$componentController`, I mean). Likewise, failing to provide a `$scope` in `locals` will throw an error in the former case (because it will try to assign the controller instance on the scope), but not in the other two cases.

File: test/ngMock/angular-mocksSpec.js (hunk 1 above)
Comment (petebacondarwin) [inline test/ngMock/angular-mocksSpec.js]: If you like but the implementation is just proxying to `$controller` anyway so it doesn't test any new paths.

File: test/ngMock/angular-mocksSpec.js (hunk 2 above)
Comment (petebacondarwin) [inline test/ngMock/angular-mocksSpec.js]: Right! Interesting that JSCS doesn't pick that up

**Comment by petebacondarwin** [comment]
@Narretz - it is much more complicated to do it for plain directives because we don't actually call the directive factory function until the directive is matched in some HTML. This means that the controller is not easily available. Components are simpler because the controller is not wrapped in a factory function.

**Comment by petebacondarwin** [comment]
>There will be some slight inconsistency between using controller: 'SomeCtrl as identifier' and controller: 'SomeCtrl' or controller: SomeCtrl. In the former case, the controller will be added as an alias on the scope, whereas in other two cases, it won't (when usign $componentController, I mean). Likewise, failing to provide a $scope in locals will throw an error in the former case (because it will try to assign the controller instance on the scope), but not in the other two cases.

@gkalpak this is also true of `ngMock.$controller` right now.

File: test/ngMock/angular-mocksSpec.js (hunk 2 above)
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: JSCS does not check for this. We'd need either the [disallowSpacesInsideObjectBrackets][1] rule or the [requireSpacesInsideObjectBrackets][2] rule, which we don't have.

I prefer disallowing spaces (which is more consistent with our disallowing spaces inside square brackets and parenthesis). Should I add a rule for object brackets ?

[1]: http://jscs.info/rule/disallowSpacesInsideObjectBrackets
[2]: http://jscs.info/rule/requireSpacesInsideObjectBrackets

**Comment by gkalpak** [comment]
@petebacondarwin, true, but things are more explicit there. You are are the one passing the argument to `$controller()` so the outcome should be less surprizing. In the case of `$componentController()` this happens indirectly; I would expect the same outcome when instantiating a component controller.

I wouldn't go to great lengths trying to "fix" this. If there's no easy fix (e.g. extracting the controller name and ignoring the alias or something), I would at least document that when using `controller: 'SomeCtrl as identifier'`, you need to supply a `$scope` in `locals`.

No biggie anyway - it only affects test code. Feel free to leave as is, if you think it's OK.
LGTM otherwise :+1:

**Comment by petebacondarwin** [comment]
@gkalpak - OK, I see what you mean. But since this is for people testing controllers that (hopefully) they have actually written, then it shouldn't be *too* surprising!

**Comment by petebacondarwin** [comment]
Actually in light of the latest discussions about `controller as` we should probably provide the `ident` param, if it is not explicitly provided, to be in line with the values of the component definition. I.E. if none is given the controller is attached to the scope via `$scope.$ctrl`.

**Comment by gkalpak** [comment]
Now it LRGTM ('R' for "really") :smiley:

**Pull request closed without merging**